MutableHttpFields.asImmutable avoids copy

gregw opened the pull request:
Avoid a copy in MutableHttpFields.asImmutable if the mutable is never mutated again.
Other minor improvements are included:
 + Better iterator usage
 + Use StringUtil.asciiEqualsIgnoreCase for `HttpHeader#is(String)`
 + Case sensitive `HttpMethod#is(String)`

gregw committed:
MutableHttpFields.asImmutable avoids copy

Avoid a copy in MutableHttpFields.asImmutable if the mutable is never mutated again.

lorban committed:
reduce instance creations needed for iterations

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix getField

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/ImmutableHttpFields.java:
```diff
@@ -59,7 +58,7 @@ public HttpFields asImmutable()
     public int hashCode()
     {
         int hash = 0;
```
shouldn't the hash function start with an arbitrary prime number? (not `0`)

lorban replied:
Agreed, since we do xor's in the loop, starting with a hash of 31 would improve the spreading.

joakime replied:
@gregw this should be a prime number.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ManagedSelector.java:
```diff
@@ -707,12 +709,12 @@ private void updateKeys()
             // This will update only those keys whose selection did not cause an
             // updateKeys update to be submitted.
-            for (SelectionKey key : _keys)
+            _keys.forEach(key ->
```
Darn, `HashSet.forEach()` is implemented with an iterator so this doesn't save any allocation.

sbordet replied:
@lorban note that `java.nio.channels.Selector` has new APIs, in particular `select(Consumer<SelectionKey>)`, which may help in not allocating the iterator.
I always wanted to look at this to see if we can improve, but never had time.

BTW, I thought these `Iterator` instances could be easily inlined.

gregw replied:
@sbordet I don't see how we can use `select(Consumer<SelectionKey>)` as we use the allocated iterator as a cursor, which each processing thread taking the next key and then handling it all the way into the application (if we have reserved threads).
With an iterator, we can have another producing thread take over the iteration, but we could not do so with the new API.

So I think the cost of that iterator allocation is some thing we have to pay for our current scheduling approach... which we should always review to check it is still good.

lorban replied:
@gregw we checked the two select implementations and they both allocate objects that escape the scope of the call, so they cannot be stack-allocated. It's unavoidable: all NIO select variants do allocate so there's no point in modifying this code.

lorban committed:
make _requestAttributeListeners list lazy

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@gregw FYI the `testBasicAuthenticationWithAuthenticationRemoved` test failures are caused by your `MutableHttpFields.asImmutable()` changes.

gregw committed:
Merge branch 'jetty-12.0.x' into experiment/12/HttpFields_asImmutable

gregw committed:
MutableHttpFields.asImmutable avoids copy

Avoid a copy in MutableHttpFields.asImmutable if the mutable is never mutated again.

gregw committed:
Fix for issue with #10563

Do not persist a defaulted charset used in the request.
Throw UnsupportedEncodingException from getReader
Improve performance with asciiEqualsIgnoreCase
HttpMethod is case-insensitive

gregw committed:
Merge branch 'jetty-12.0.x' into experiment/12/HttpFields_asImmutable

lorban commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpMethod.java:
```diff
@@ -95,7 +95,7 @@ public byte[] getBytes()
     public boolean is(String s)
     {
-        return toString().equalsIgnoreCase(s);
+        return toString().equals(s);
```
I think this is more spec-compliant, but this is a behavioral change from versions 9.4/10/11. I'm worried this could cause issues with users unknowingly relying on this case-insensitive comparison down the road.

gregw replied:
Actually I don't think this actually does change behaviour in the common cases.  We always lookup a method in the cache (even when in case insensitive mode) and replace the string with the one we find.  Thus any code that does `HttpMethod.FOOis(request.getMethod())` will see the "normalized" case method name.     Thus I'm comfortable making this change and dealing with any consequences.

sbordet replied:
A nit, this should be `name().equals(s)`.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/StringUtil.java:
```diff
@@ -225,18 +225,71 @@ public static String sanitizeFileSystemName(String str)
+     * @return true if the string ends with the substring, ignoring {@link StandardCharsets#US_ASCII}  case differences.
+     * @deprecated Use {@link #asciiEndsWithIgnoreCase(String, String)}
+     */
+    @Deprecated
```
`@Deprecated(since = "12.0.3")` ?

sbordet replied:
I'm not against the `since` attribute, but either we are consistent, run through the code add it to all deprecations, and we have a policy, otherwise it's not so useful and inconsistent.

lorban committed:
make violations set lazy + add a quicker way to allocate an empty violations EnumSet

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
revert iteration change in updateKeys() as that doesn't save any allocation

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Merge branch 'jetty-12.0.x' into experiment/12/HttpFields_asImmutable

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/StringUtil.java:
```diff
@@ -225,18 +225,71 @@ public static String sanitizeFileSystemName(String str)
             return true;
 
-        if (s == null || s.length() < w.length())
+        if (string == null || string.length() < prefix.length())
```
Another quick check: If the `prefix` is zero length, always return true.

joakime commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/StringUtil.java:
```diff
@@ -250,23 +303,42 @@ public static boolean startsWithIgnoreCase(String s, String w)
-        int sl = s.length();
-        int wl = w.length();
+        int sl = string.length();
+        int wl = suffix.length();
```
Another quick check: If the `wl` is 0 (empty string), always return true.

joakime reviewed (commented):
Some open thoughts

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MutableHttpFields.java:
```diff
@@ -67,7 +68,17 @@ protected MutableHttpFields(int capacity)
+        }
+        else
+        {
+            _fields = new HttpField[INITIAL_SIZE];
```
You know the size of the fields, so you may use that instead of `INITIAL_SIZE`, no?

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -170,6 +170,14 @@ public String getDescription()
     private static final List<UriCompliance> KNOWN_MODES = List.of(DEFAULT, LEGACY, RFC3986, UNAMBIGUOUS, UNSAFE);
+    private static final EnumSet<Violation> NO_VIOLATION = EnumSet.noneOf(Violation.class);
+
+    // This saves a AccessController.executePrivileged() call caused by
```
`executePrivileged` -> `doPrivileged`.

gregw replied:
I'm going to make the constant a `Set<Violation>` as then I can make it immutable and we can avoid even the simple clone operation.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/UriCompliance.java:
```diff
@@ -170,6 +170,14 @@ public String getDescription()
+
+    // This saves a AccessController.executePrivileged() call caused by
+    // EnumSet.noneOf() calling EnumSet.getUniverse().
+    public static EnumSet<Violation> emptyViolationsEnumSet()
```
I would call this method `newViolationsEnumSet()` because the prefix `empty` typically refers to a constant, which this is not (for example `Optional.empty()`, `Collections.empty*()`, etc.).

gregw replied:
I've removed this method altogether.

gregw committed:
updates from review

joakime reviewed (dismissed):
The `hashCode()` implementation in `ImmutableHttpFields` is non-ideal.
But that hints at a larger issue in the codebase.
Probably should review that across the codebase in a future PR.

gregw committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into experiment/12/HttpFields_asImmutable

gregw committed:
updates from review

prime hash seeds

gregw commented:
@joakime this is already a bit of an omnibus PR, so I have fixed the other hash calculations I can see that are based on 0.

lorban committed:
simplify HttpStreamOverHTTP1 ids

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Disable thread-local cache

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/MutableHttpFields.java:
```diff
@@ -67,7 +68,21 @@ protected MutableHttpFields(int capacity)
+        }
+        else if (fields != null)
+        {
+            _fields = new HttpField[fields.size() + INITIAL_SIZE];
```
```suggestion
            _fields = new HttpField[fields.size() + SIZE_INCREMENT];
```

lorban commented:
@gregw I have an extra set of small but potentially controversial changes worth 1-2% of perf in my `perf/jetty-12-HttpURI-violations` branch. Would you like me to push them here?

gregw committed:
updates from review

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -479,10 +479,10 @@ private static class RetainedBucket
         {
             if (poolSize <= ConcurrentPool.OPTIMAL_MAX_SIZE)
-                _pool = new ConcurrentPool<>(ConcurrentPool.StrategyType.THREAD_ID, poolSize, true);
+                _pool = new ConcurrentPool<>(ConcurrentPool.StrategyType.THREAD_ID, poolSize, false);
```
@lorban I like this change as THREAD_ID and the ThreadLocal cache are kind of trying to do the same thing anyway.
However, I wonder if we could improve the THREAD_ID algorithm a little bit.  Currently it is used as:
```
            case THREAD_ID -> (int)(Thread.currentThread().getId() % size);
```

The problem here is that typically the threads in use will be consecutive IDs, so if there are 50 threads in use and the pool is 100 in size, it may be that 50 consecutive entries are all in use and that an acquire by a thread that already has used one buffer will now need to scan over 49 others to find an empty slot.  Perhaps we should be treating it a little more like an initial hash and spreading the thread ID over the available size more.    Perhaps the following would work a little bit better:

```
            case THREAD_ID -> (int)((Thread.currentThread().getId() * 31) % size);
```

(31 selected because it is prime)

lorban replied:
I modified the THREAD_ID strategy to multiply by 31, ran the benchmark, and measured no noticeable change.

Using the thread id as an index combined with the randomness of the OS' scheduler is already spreading the threads sufficiently well, it seems.

gregw replied:
Still it might be best to keep the * 31.   It is probably rare that all threads need buffers at exactly the same time, but I also think it could occur.  It may be the benchmark is not the right pattern to make it happen.   So you saw no cost to doing this, so let's keep it.

lorban replied:
I wrote a micro benchmark to see exactly what's going on and the * 31 costs exactly two cheap instructions: shift left by 5, then subtract once.

No big deal, so I'm okay to keep this change.

lorban committed:
multiply thread ID by a prime to make spreading more random

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
@sbordet nudge

gregw committed:
Merge branch 'jetty-12.0.x' into experiment/12/HttpFields_asImmutable
